node/src/components/block_validator.rs

```diff
@@ -216,9 +217,98 @@ where
+        }
+        effects
+    }
+}
 
-                effects
+/// Block validator.
+#[derive(Debug)]
+pub(crate) enum BlockValidatorState<T, I> {
+    Loading(Vec<Event<T, I>>),
+    Ready(BlockValidatorReady<T, I>),
+}
+
+/// Block validator.
+#[derive(DataSize, Debug)]
+pub(crate) struct BlockValidator<T, I> {
+    #[data_size(skip)]
+    state: BlockValidatorState<T, I>,
+    _marker: std::marker::PhantomData<I>,
+}
+
+impl<T, I> BlockValidator<T, I>
+where
+    T: BlockLike + Send + Clone + 'static,
+    I: Clone + Send + 'static + Send,
+{
+    /// Creates a new block validator instance.
+    pub(crate) fn new<REv>(effect_builder: EffectBuilder<REv>) -> (Self, Effects<Event<T, I>>)
+    where
+        REv: From<Event<T, I>> + From<StorageRequest> + Send + 'static,
+    {
+        let effects = async move {
+            effect_builder
+                .get_chainspec(Version::new(1, 0, 0))
+                .await
+                .expect("chainspec should be infallible")
+        }
+        .event(move |chainspec| Event::Loaded { chainspec });
+        (
+            BlockValidator {
+                state: BlockValidatorState::Loading(Vec::new()),
+                _marker: std::marker::PhantomData,
+            },
+            effects,
+        )
+    }
+}
+
+impl<T, I, REv> Component<REv> for BlockValidator<T, I>
+where
+    T: BlockLike + Send + Clone + 'static,
+    I: Clone + Send + 'static,
+    REv: From<Event<T, I>>
+        + From<BlockValidationRequest<T, I>>
+        + From<FetcherRequest<I, Deploy>>
+        + From<StorageRequest>
+        + Send,
+{
+    type Event = Event<T, I>;
+    type ConstructionError = Infallible;
+
+    fn handle_event(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        _rng: &mut NodeRng,
+        event: Self::Event,
+    ) -> Effects<Self::Event> {
+        let mut effects = Effects::new();
+        match (&mut self.state, event) {
+            (BlockValidatorState::Loading(requests), Event::Loaded { chainspec }) => {
+                let mut new_ready_state = BlockValidatorReady {
+                    chainspec,
+                    validation_states: Default::default(),
+                    in_flight: Default::default(),
+                    _marker: std::marker::PhantomData,
+                };
+                let requests = requests;
+                // Replay postponed events onto new state.
+                for ev in requests.drain(..) {
+                    effects.extend(new_ready_state.handle_event(effect_builder, ev));
+                }
+                self.state = BlockValidatorState::Ready(new_ready_state);
+            }
+            (BlockValidatorState::Loading(requests), request @ Event::Request(_)) => {
+                requests.push(request);
+            }
+            (BlockValidatorState::Ready(ref mut ready_state), event) => {
+                effects.extend(ready_state.handle_event(effect_builder, event));
             }
+            _ => {}
```

Comment:
I think we should at least log an error here - I know we shouldn't get to this branch, but still, if we do, it's a bug (we'd have got a `DeployFound` or `DeployMissing` event while in the loading state).

Reply:
Made this more explicit as to what match arm was being hit, now logging an error.